orca/service.go

```diff
@@ -0,0 +1,214 @@
+	// metrics are reported on the streaming RPC initiated by the client. If
+	// unspecified, negative or less than the default value of 30s, the default
+	// is used. Clients may request a higher value as part of the
+	// StreamCoreMetrics streaming RPC.
+	MinReportingInterval time.Duration
+
+	// Allow a minReportingInterval which is less than the default of 30s.
+	// Used for testing purposes only.
+	allowAnyMinReportingInterval bool
+}
+
+// Register creates a new ORCA service implementation configured using the
+// provided options and registers the same on the provided service registrar.
+func Register(s grpc.ServiceRegistrar, opts ServiceOptions) (*Service, error) {
+	// TODO(easwars): Once the generated pb.gos are updated, we wont need this
+	// type assertion any more, since the new versions accept a
+	// grpc.ServiceRegistrar in their registration functions.
+	srv, ok := s.(*grpc.Server)
+	if !ok {
+		return nil, fmt.Errorf("concrete type of provided grpc.ServiceRegistrar is %T, only supported type is %T", s, &grpc.Server{})
+	}
+
+	// The default minimum supported reporting interval value can be overridden
+	// for testing purposes through the orca internal package.
+	if !opts.allowAnyMinReportingInterval {
+		if opts.MinReportingInterval < 0 || opts.MinReportingInterval < minReportingInterval {
+			opts.MinReportingInterval = minReportingInterval
+		}
+	}
+	service := &Service{
+		minReportingInterval: opts.MinReportingInterval,
+		utilization:          make(map[string]float64),
+	}
+	v3orcaservicegrpc.RegisterOpenRcaServiceServer(srv, service)
+	return service, nil
+}
+
+// determineReportingInterval determines the reporting interval for out-of-band
+// metrics. If the reporting interval is not specified in the request, or is
+// negative or is less than the configured minimum (via
+// ServiceOptions.MinReportingInterval), the latter is used. Else the value from
+// the incoming request is used.
+func (s *Service) determineReportingInterval(req *v3orcaservicegrpc.OrcaLoadReportRequest) time.Duration {
+	if req.GetReportInterval() == nil {
+		return s.minReportingInterval
+	}
+	dur := req.GetReportInterval().AsDuration()
+	if dur < 0 || dur < s.minReportingInterval {
+		logger.Warningf("Received reporting interval %q is less than configured minimum: %v. Using default: %s", dur, s.minReportingInterval)
+		return s.minReportingInterval
+	}
+	return dur
+}
+
+func (s *Service) sendMetricsResponse(stream v3orcaservicegrpc.OpenRcaService_StreamCoreMetricsServer) error {
+	return stream.Send(s.toLoadReportProto())
+}
+
+// StreamCoreMetrics streams custom backend metrics injected by the server
+// application.
+func (s *Service) StreamCoreMetrics(req *v3orcaservicepb.OrcaLoadReportRequest, stream v3orcaservicepb.OpenRcaService_StreamCoreMetricsServer) error {
+	// Send one response immediately before creating and waiting for the timer
+	// to fire.
+	if err := s.sendMetricsResponse(stream); err != nil {
+		return err
+	}
```

Comment:
Optional: you could remove this is you do it at the top of the `for` loop instead of when the ticker fires.

Reply:
Done.